File: OctopusNode/src/main/java/com/octopus/node/service/JobExecutors.java
package com.octopus.node.service;

import com.octopus.node.core.JobExecutor;
import com.octopus.node.core.JobType;
import com.octopus.transport.SubJobResult;
import com.octopus.transport.SubJob;
import org.apache.log4j.Logger;

import java.util.HashMap;
import java.util.Map;

public class JobExecutors {
    private final Map<JobType, JobExecutor> executors = new HashMap<>();

    // Used for executors not yet implemented to avoid a null pointer exception
    private static class NullJobExecutor implements JobExecutor {
        private final Logger LOG = Logger.getLogger(NullJobExecutor.class);

        @Override
        public SubJobResult execute(SubJob subJob) {
            LOG.warn("The null job executor was called.");
            return new SubJobResult(0,null, JobType.NONE, "", 0);
        }

        @Override
        public JobType getJobType() {
            return JobType.NONE;
        }
    }

    /**
     * All new {@link JobExecutor}s must be added here.
     */
    public JobExecutors() {
        executors.put(JobType.WEB_SCRAPING, new WebScrapingJobExecutor());
        executors.put(JobType.VIDEO_TRANSCODE, new VideoJobExecutor());
    }

    public JobExecutor get(JobType type) {
        return executors.getOrDefault(type, new NullJobExecutor());
    }
}

File: OctopusService/src/main/kotlin/com/octopus/service/JobExecutors.kt
package com.octopus.service

import com.octopus.core.enums.JobType
import com.octopus.dao.JobDao
import com.octopus.dao.SubJobDao
import com.octopus.service.video.VideoCollector
import com.octopus.service.video.VideoExecutorService
import com.octopus.service.video.VideoMakerService
import com.octopus.service.webscraping.WebScrapingCollector
import com.octopus.service.webscraping.WebScrapingExecutorService
import com.octopus.service.webscraping.WebScrapingMakerService
import java.util.*

class JobExecutors(jobDao: JobDao, subJobDao: SubJobDao) {
    private val executors: MutableMap<JobType, JobExecutor> = EnumMap(JobType::class.java)

    // To avoid creating a new NullJobExecutor each call
    companion object {
        private val nullJobExecutor = NullJobExecutor()
    }

    fun get(type: JobType): JobExecutor {
        return executors.getOrDefault(type, nullJobExecutor)
    }

    /**
     * All new [JobExecutor]s must be added here.
     */
    init {
        // Add WebScraping Executor
        executors[JobType.WEB_SCRAPING] = WebScrapingExecutorService(
                WebScrapingMakerService(jobDao, subJobDao),
                WebScrapingCollector(jobDao, subJobDao),
                subJobDao, jobDao
        )

        // Add Video Executor
        executors[JobType.VIDEO_TRANSCODE] = VideoExecutorService(
                VideoMakerService(jobDao, subJobDao),
                VideoCollector(jobDao, subJobDao),
                subJobDao, jobDao
        )
    }
}
